src/queryIterator.ts

```diff
@@ -174,13 +177,41 @@ export class QueryIterator<T> {
     );
   }
 
-  private createQueryExecutionContext() {
-    return new ProxyQueryExecutionContext(
+  private createPipelinedExecutionContext(partitionedExecutionInfo: PartitionedQueryExecutionInfo) {
+    return new PipelinedQueryExecutionContext(
       this.clientContext,
+      this.resourceLink,
       this.query,
       this.options,
-      this.fetchFunctions,
-      this.resourceLink
+      partitionedExecutionInfo
     );
   }
+
+  private isPartitionedExecutionError(error: any): error is ErrorResponse {
+    return (
+      error.code === StatusCodes.BadRequest &&
+      "substatus" in error &&
+      error["substatus"] === SubStatusCodes.CrossPartitionQueryNotServable
+    );
+  }
+
+  private async executeMethod(methodName: Exclude<keyof IExecutionContext, "hasMoreResults">) {
+    try {
+      return await this.queryExecutionContext[methodName]();
```

Comment:
can we do a callback function instead?

Reply:
Trying to avoid callbacks in favor of async/await